### PR TITLE
feat(gtest): Implement send_signal and system gas reserve/unreserve

### PR DESCRIPTION
**Pull request by playX18**
- Implemented `send_signal` and `system_reserve_gas`, `system_unreserve_gas`.
- Added helper methods to Gas tree manager to reserve/unreserve system gas.

@techraed

### CODE REVIEW

File: gtest/src/program.rs [hunk 1]
```diff
@@ -884,6 +884,24 @@ mod tests {
...
+
+        let prog = Program::from_binary_with_id(&sys, 137, signal_in_handle::WASM_BINARY);
+
+        let run_result = prog.send_bytes(user_id, b"Does not matter");
+
+        assert!(!run_result.main_failed());
+
+        let run_result = prog.send(user_id, String::from("should panic"));
+        log::debug!("{:?}", run_result);
+        run_result.assert_panicked_with("Gotcha!");
```
Comment (techraed) [inline gtest/src/program.rs:902]: Your test program actually sends a message `handle` from program to user, which must go to mailbox.
I'd adjust the test:
1. include some payload to the message (on the contract)
2. check here that there's a message in the mailbox with the same payload. 

Currently test doesn't actually check that signals work on gtest.

**Review by techraed — CHANGES_REQUESTED** [review]
Instead of creating another one wasm contract in `examples` I'd add system reservation functionality into `demo_constructor` and use it to create a desired logic instead another crate in the workspace.

File: gtest/src/program.rs [hunk 2]
```diff
@@ -880,26 +880,41 @@ pub mod gbuild {
...
-        let prog = Program::from_binary_with_id(&sys, 137, signal_in_handle::WASM_BINARY);
+        let scheme = Scheme::predefined(
+            Calls::builder().noop(),
+            Calls::builder()
+                .system_reserve_gas(1_000_000_000)
+                .panic("Gotcha!"),
+            Calls::builder().noop(),
+            Calls::builder().send(
+                Arg::new(ProgramIdWrapper::from(user_id).0.into_bytes()),
+                Arg::bytes("Signal handled"),
```
Comment (techraed) [inline gtest/src/program.rs:904]: please create a separate variable for the `Signal handled` string.

**Review by techraed — APPROVED** [review]

File: gtest/src/program.rs (hunk 2 above)
Comment (techraed) [reply gtest/src/program.rs:904]: as it is re-used

**Comment by techraed** [comment]
@playX18 resolve conflict with master and merge it, please

File: Cargo.toml [hunk 3]
```diff
@@ -301,7 +301,9 @@ sandbox-wasmer = { package = "wasmer", version = "2.2", features = [
     "singlepass",
 ] }
 sandbox-wasmer-types = { package = "wasmer-types", version = "2.2" }
-sandbox-wasmi = { package = "wasmi", git = "https://github.com/gear-tech/wasmi", branch = "v0.13.2-sign-ext", features = ["virtual_memory"] }
+sandbox-wasmi = { package = "wasmi", git = "https://github.com/gear-tech/wasmi", branch = "v0.13.2-sign-ext", features = [
```
Comment (breathx) [inline Cargo.toml:304]: I don't really like such formatting for features

Comment (breathx) [reply Cargo.toml:304]: And everywhere else its stated differently

File: gtest/src/manager.rs [hunk 4]
```diff
@@ -1240,11 +1240,50 @@ impl JournalHandler for ExtManager {
...
+                .split_with_value(
+                    trap_signal.is_reply(),
+                    message_id,
+                    trap_signal.id(),
+                    reserved,
+                )
+                .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
 
-    fn send_signal(&mut self, _message_id: MessageId, _destination: ProgramId, _code: SignalCode) {}
+            self.dispatches.push_front(trap_signal);
```
Comment (breathx) [inline gtest/src/manager.rs:1282]: why front?

**Review by breathx — CHANGES_REQUESTED** [review]

**Comment by breathx** [comment]
Overall looks great. Push back into queue and resolve conflicts please

**Review by breathx — APPROVED** [review]
Still need to merge master

**Pull request merged**